plutus-tx-plugin/src/PlutusTx/Options.hs

```diff
@@ -0,0 +1,234 @@
+        , let k = "profile-all"
+           in (k, PluginOption typeRep (flag (const All) k) poProfile showText mempty)
+        , let k = "coverage-all"
+           in (k, PluginOption typeRep (setTrue k) poCoverageAll showText mempty)
+        , let k = "coverage-location"
+           in (k, PluginOption typeRep (setTrue k) poCoverageLocation showText mempty)
+        , let k = "coverage-boolean"
+           in (k, PluginOption typeRep (setTrue k) poCoverageBoolean showText mempty)
+        , let k = "remove-trace"
+           in (k, PluginOption typeRep (setTrue k) poRemoveTrace showText mempty)
+        ]
+  where
+    showText :: Show a => a -> Text
+    showText = Text.pack . show
+
+flag :: (a -> a) -> OptionKey -> Maybe OptionValue -> Validation ParseError (a -> a)
+flag f k = maybe (Success f) (Failure . UnexpectedValue k)
+
+setTrue :: OptionKey -> Maybe OptionValue -> Validation ParseError (Bool -> Bool)
+setTrue = flag (const True)
+
+setFalse :: OptionKey -> Maybe OptionValue -> Validation ParseError (Bool -> Bool)
+setFalse = flag (const False)
+
+intOption :: OptionKey -> Maybe OptionValue -> Validation ParseError (Int -> Int)
+intOption k = \case
+    Just v ->
+        maybe
+            (Failure (CannotParseValue k v (someTypeRep (Proxy @Int))))
+            (Success . const)
+            (readMaybe (Text.unpack v))
+    Nothing -> Failure (MissingValue k)
+
+defaultPluginOptions :: PluginOptions
+defaultPluginOptions =
+    PluginOptions
+        { _poDoTypecheck = True
+        , _poDeferErrors = False
+        , _poContextLevel = 1
+        , _poDumpPir = False
+        , _poDumpPlc = False
+        , _poDumpUPlc = False
+        , _poOptimize = True
+        , _poPedantic = False
+        , _poVerbose = False
+        , _poDebug = False
+        , _poMaxSimplifierIterations = view PIR.coMaxSimplifierIterations PIR.defaultCompilationOpts
+        , _poDoSimplifierUnwrapCancel = True
+        , _poDoSimplifierBeta = True
+        , _poDoSimplifierInline = True
+        , _poDoSimplifierRemoveDeadBindings = True
+        , _poProfile = None
+        , _poCoverageAll = False
+        , _poCoverageLocation = False
+        , _poCoverageBoolean = False
+        , _poRemoveTrace = False
+        }
+
+processOne ::
+    OptionKey ->
+    Maybe OptionValue ->
+    Validation ParseError (PluginOptions -> PluginOptions)
+processOne key val = case Map.lookup key pluginOptions of
+    Just (PluginOption _ f field _ _) -> over field <$> f val
+    Nothing ->
+        let suggs =
+                Text.pack
+                    <$> GHC.fuzzyMatch (Text.unpack key) (Text.unpack <$> Map.keys pluginOptions)
+         in Failure (UnrecognisedOption key suggs)
+
+processAll ::
+    [(OptionKey, Maybe OptionValue)] ->
+    Validation ParseErrors [PluginOptions -> PluginOptions]
+processAll = traverse $ first (ParseErrors . pure) . uncurry processOne
```

Comment:
For example, here's how I'd do this one:
```
processAll kvs = for kvs $ \(k, v) -> 
    let res = processOne k v
    -- inject the individual ParseError into ParseErrors
    in first (ParseErrors . pure) res
```

Reply:
FWIW @zliu41's version is more readable to me than @michaelpj's one. I see `Validation` and `traverse` -- and I already know what's happening with errors. `first (ParseErrors . pure)` is trivial noise, so documenting it only creates more noise.

In this example there isn't much difference, but if it was bigger, then binding `k` and `v` explicitly would result in me checking the rest of the function to see if they are used anywhere apart from `processOne`. The more variables there are, the more non-trivial relationships between them there can be, the more vigilant I need to be. Pointfree definitions that only contain `$` and `.` (without any `(f .)` etc) are much more readable to me than loads of explicit bindings and comments about noise.

YMMV

Reply:
Yeah I agree with @effectfully on this one. Someone reading the code should merely see it as `processAll = traverse processOne` and ignore the rest. I'll tweak the other one (`toKeyValue`) to make it slightly more readable.

Reply:
Right, so that would also be nice I think:
```
processAll = traverse processOne'
  where
     -- Adapt to consume the tuples and inject the errors into ParseErrors
     processOne' (k, v) = first (ParseErrors . pure) $ processOne k v
```